File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/heroes_list_component.dart
// #docregion
import 'package:angular2/core.dart';

import 'edit_item.dart';
import 'hero.dart';
import 'hero_card_component.dart';
import 'hero_editor_component.dart';
import 'heroes_service.dart';

@Component(
    selector: 'heroes-list',
    template: '''
  <div>
      <ul>
        <li *ngFor="let editItem of heroes">
          <hero-card
            [hidden]="editItem.editing"
            [hero]="editItem.item">
          </hero-card>
          <button
            [hidden]="editItem.editing"
            (click)="editItem.editing = true">
              edit
          </button>
          <hero-editor
            (saved)="onSaved(editItem, \$event)"
            (canceled)="onCanceled(editItem)"
            [hidden]="!editItem.editing"
            [hero]="editItem.item">
          </hero-editor>
        </li>
      </ul>
    </div>
  ''',
    directives: const [HeroCardComponent, HeroEditorComponent])
class HeroesListComponent {
  List<EditItem<Hero>> heroes;
  HeroesListComponent(HeroesService heroesService) {
    heroes = heroesService
        .getHeroes()
        .map((Hero item) => new EditItem(item))
        .toList();
  }

  onCanceled(EditItem<Hero> editItem) {
    editItem.editing = false;
  }

  onSaved(EditItem<Hero> editItem, Hero updatedHero) {
    editItem.item = updatedHero;
    editItem.editing = false;
  }
}
// #enddocregion

File: public/docs/_examples/hierarchical-dependency-injection/ts/app/heroes-list.component.ts
// #docregion
import { Component }  from '@angular/core';
import { Observable } from 'rxjs/Observable';

import { Hero }          from './hero';
import { HeroesService } from './heroes.service';

@Component({
  selector: 'heroes-list',
  template: `
    <div>
      <h3>Heroes</h3>
      <ul>
        <li *ngFor="let hero of heroes | async" (click)="addCard(hero)">
          {{hero.name}} ({{hero.power}})
        </li>
      </ul>
      <hero-card
        *ngFor="let selectedHero of selectedHeroes; let i = index"
        [hero]="selectedHero"
        (close)="closeCard(i)">
      </hero-card>
    </div>
    `,
  styles: [ 'li {cursor: pointer;}' ]
})
export class HeroesListComponent {
  heroes: Observable<Hero[]>;
  selectedHeroes: Hero[] = [];

  constructor(private heroesService: HeroesService) {
    this.heroes = heroesService.getHeroes();
  }

  addCard(hero: Hero) {
    this.selectedHeroes.push(hero);
  }

  closeCard(ix: number) {
    this.selectedHeroes.splice(ix, 1);
  }
}


// #enddocregion
